packages/hash/api/src/model/util.ts

```diff
@@ -305,44 +311,63 @@ export const generateWorkspaceEntityTypeSchema = (
 
   /** @todo - clean this up to be more readable: https://app.asana.com/0/1202805690238892/1202931031833226/f */
   const properties = params.properties.reduce(
-    (prev, { propertyTypeId, propertyTypeBaseUri, array }) => ({
+    (prev, { propertyTypeModel, array }) => ({
       ...prev,
-      [propertyTypeBaseUri]: array
+      [propertyTypeModel.baseUri]: array
         ? {
             type: "array",
-            items: { $ref: propertyTypeId },
+            items: { $ref: propertyTypeModel.schema.$id },
             ...(array === true ? {} : array),
           }
-        : { $ref: propertyTypeId },
+        : { $ref: propertyTypeModel.schema.$id },
     }),
     {},
   );
 
   const requiredProperties = params.properties
     .filter(({ required }) => !!required)
-    .map(({ propertyTypeBaseUri }) => propertyTypeBaseUri);
+    .map(({ propertyTypeModel }) => propertyTypeModel.baseUri);
 
   const links =
     params.outgoingLinks.length > 0
       ? params.outgoingLinks.reduce<EntityType["links"]>(
-          (prev, { linkTypeId, destinationEntityTypeId, array, ordered }) => ({
-            ...prev,
-            [linkTypeId]: array
-              ? {
-                  type: "array",
-                  items: { $ref: destinationEntityTypeId },
-                  ordered,
-                  ...(array === true ? {} : array),
-                }
-              : { $ref: destinationEntityTypeId },
-          }),
+          (
+            prev,
+            { linkTypeModel, destinationEntityTypeModels, array, ordered },
+          ) => {
+            const oneOf = {
+              oneOf: destinationEntityTypeModels.map(
+                (entityTypeModelOrReference) => {
+                  let referenceId;
+                  if (entityTypeModelOrReference === SELF_REFERENCE_MARKER) {
+                    referenceId = $id;
+                  } else {
+                    referenceId = entityTypeModelOrReference.schema.$id;
+                  }
+                  return { $ref: referenceId };
+                },
```

Comment:
Nit:
```suggestion
                (entityTypeModelOrReference) => ({
                  $ref:
                    entityTypeModelOrReference === SELF_REFERENCE_MARKER
                      ? $id
                      : entityTypeModelOrReference.schema.$id,
                }),
```

Reply:
Not 100% on this one because I like the explicit return being clear that the object is of the form `{ $ref: ID }` **_but_** I also like the removed repetition and such in the suggestion. I've added it and will let the style mull over in my head 😁 

Thanks!